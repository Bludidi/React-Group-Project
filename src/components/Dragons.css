.dragons-container > * {
  margin-bottom: 25px;
}

.dragon {
  display: flex;
  align-items: initial;
  gap: 20px;
}

.dragon-image {
  width: 20%;
  height: 20%;
}

.dragon-details > * {
  margin-bottom: 15px;
}

.dragon-name {
  font-size: 25px;
}

.dragon-description {
  font-size: 15px;
}

.dragon-button {
  color: #fff;
  background-color: #4e4eff;
  padding: 15px;
  border: 0;
  border-radius: 5px;
  cursor: pointer;
}

.reserved-btn {
  padding: 5px 10px;
  color: #fff;
  background-color: #24a0b6;
  font-weight: 800;
  border: 0;
  border-radius: 5px;
  margin-right: 10px;
}

.cancel-dragon-button {
  color: #8e8e8e;
  background-color: #fff;
  border: #717171 2px solid;
  border-radius: 5px;
  font-size: 20px;
  padding: 10px;
  cursor: pointer;
}
