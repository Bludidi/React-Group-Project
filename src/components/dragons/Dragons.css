.dragons-container > * {
  margin-bottom: 25px;
  list-style: none;
}

.dragon {
  display: flex;
  align-items: initial;
  gap: 20px;
}

.dragon-image {
  width: 20%;
  height: 20%;
}

.dragon-details > * {
  margin-bottom: 15px;
}

.dragon-name {
  font-size: 25px;
}

.dragon-description {
  font-size: 15px;
}

.dragon-button {
  color: #fff;
  background-color: #1083ff;
  padding: 15px;
  border: 0;
  border-radius: 5px;
  cursor: pointer;
  font-size: 20px;
}

.reserved-btn {
  padding: 5px 10px;
  color: #fff;
  background-color: #24a0b6;
  font-weight: 600;
  border: 0;
  border-radius: 5px;
  margin-right: 10px;
}

.cancel-dragon-button {
  color: #8e8e8e;
  background-color: #fff;
  border: #717171 2px solid;
  border-radius: 5px;
  font-size: 20px;
  padding: 10px;
  cursor: pointer;
}

.main-container {
  display: flex;
  gap: 15px;
}

.reserved-dragons {
  min-width: 400px;
}

.dragons-cont {
  border: #c4c4c4 2px solid;
  border-radius: 8px;
}

.dragons-tit {
  padding: 10px;
}

.dragon-n {
  padding: 15px;
  border-bottom: #c4c4c4 1px solid;
  list-style: none;
  font-size: 18px;
  font-weight: 700;
}