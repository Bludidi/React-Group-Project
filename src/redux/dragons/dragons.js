import axios from 'axios';
import produce from 'immer';

const url = 'https://api.spacexdata.com/v3/dragons';
const initialState = [];

const fetchDragons = async () => {
  const response = await axios.get(url);
  const dragonsData = response.data;
  return dragonsData;
};

const GET_DRAGONS = 'GET_DRAGONS';
const BOOK_DRAGON = 'BOOK_DRAGON';
const CANCEL_BOOKING = 'CANCEL_BOOKING';

const getDragons = (payload) => ({
  type: GET_DRAGONS,
  payload,
});

export const bookDragon = (payload) => ({
  type: BOOK_DRAGON,
  payload,
});

export const cancelBooking = (payload) => ({
  type: CANCEL_BOOKING,
  payload,
})

const dragonsReducer = (state = initialState, { type, payload }) => {
  let newState = [];
  switch (type) {
    case GET_DRAGONS:
      newState = produce(state, (draftState) => Object.keys(payload).forEach((key) => {
        draftState.push({
          id: payload[key].id,
          name: payload[key].name,
          description: payload[key].description,
          flickr_images: payload[key].flickr_images[0],
          reserved: false,
        });
      }));
      return newState;
    case BOOK_DRAGON:
      newState = state.map((dragon) => {
        if (dragon.id === payload) return { ...dragon, reserved: true };
        return dragon;
      });
      return newState;
    case CANCEL_BOOKING:
      newState = state.map((dragon) => {
        if (dragon.id === payload) return { ...dragon, reserved: false };
        return dragon;
      });
      return newState;
    default:
      return state;
  }
};

export function getAllDragons() {
  return async function (dispatch) {
    const allDragons = await fetchDragons();
    (dispatch(getDragons(allDragons)));
  };
}

export default dragonsReducer;
