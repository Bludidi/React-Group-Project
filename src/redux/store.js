import { configureStore } from '@reduxjs/toolkit';
import missionReducer from './missions/mission.redux';

export default configureStore({
  reducer: {
    mission: missionReducer,
  },
});
