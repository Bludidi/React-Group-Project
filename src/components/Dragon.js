import React from 'react';
import { useDispatch } from 'react-redux';
import { bookDragon, cancelBooking } from '../redux/dragons/dragons';

function Dragon(props) {
  const dragon = props;
  const dispatch = useDispatch();

  const bookADragon = (id) => {
    dispatch(bookDragon(id));
  };

  const cancelDragonBooking = (id) => {
    dispatch(cancelBooking(id));
  };

  return dragon.reserved ? (
    <div className="dragon">
      <img src={dragon.image} alt="Dragon" className="dragon-image" />
      <div className="dragon-details">
        <h3 className="dragon-name">{dragon.name}</h3>
        <p className="dragon-description">
          <button type="button" className="reserved-btn">Reserved</button>
          {dragon.description}
        </p>
        <button
          className="cancel-dragon-button"
          onClick={() => cancelDragonBooking(dragon.id)}
          type="button"
        >
          Cancel Reservation
        </button>
      </div>
    </div>
  ) : (
    <div className="dragon">
      <img src={dragon.image} alt="Dragon" className="dragon-image" />
      <div className="dragon-details">
        <h3 className="dragon-name">{dragon.name}</h3>
        <p className="dragon-description">{dragon.description}</p>
        <button
          className="dragon-button"
          onClick={() => bookADragon(dragon.id)}
          type="button"
        >
          Reserve Dragon
        </button>
      </div>
    </div>
  );
}

export default Dragon;
