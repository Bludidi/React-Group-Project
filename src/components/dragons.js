import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Dragon from './Dragon';
import { getAllDragons } from '../redux/dragons/dragons';
import './Dragons.css';

function Dragons() {
  const dispatch = useDispatch();
  const dragons = useSelector((state) => state.dragons);

  useEffect(() => {
    if (dragons.length === 0) {
      dispatch(getAllDragons());
    }
  }, []);

  return (
    <div>
      <ul className="dragons-container">
        {
          dragons.map((dragon) => (
            <li key={dragon.id}>
              <Dragon
                id={dragon.id}
                description={dragon.description}
                name={dragon.name}
                image={dragon.flickr_images}
              />
            </li>
          ))
        }
      </ul>
    </div>
  );
}

export default Dragons;
