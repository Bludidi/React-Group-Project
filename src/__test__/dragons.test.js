import { render } from "@testing-library/react";
import React from 'react';
import { Provider } from 'react-redux';
import store from '../redux/configureStore';
import Dragon from "../components/dragons/Dragon";
import Dragons from "../components/dragons/dragons";
import { bookDragon, cancelBooking, getDragons } from "../redux/dragons/dragons";

describe('Dragons', () => {
  it('Dragon component is rendered correctly', () => {
    const dragon = render(
      <Provider store={store}>
        <Dragon />
      </Provider>,
    );
    expect(dragon).toMatchSnapshot();
  });

  it('Dragons component is rendered correctly', () => {
    const dragons = render(
      <Provider store={store}>
        <Dragons />
      </Provider>,
    );
    expect(dragons).toMatchSnapshot();
  });

  it('should return an action with type GET_DRAGONS', () => {
    const data = [
      {
        id: 1,
        name: 'Dragon1',
        description: 'description of the first one',
        flickr_images: [],
      },
    ];
    const action = getDragons(data);
    expect(action.type).toBe('GET_DRAGONS');
  });

  it('should return an action with type BOOK_DRAGON', () => {
    const id = 1;
    const action = bookDragon(id);
    expect(action.type).toBe('BOOK_DRAGON');
  });

  it('should should return an action with type CANCEL_BOOKING', () => {
    const id = 1;
    const action = cancelBooking(id);
    expect(action.type).toBe('CANCEL_BOOKING');
  });
});
