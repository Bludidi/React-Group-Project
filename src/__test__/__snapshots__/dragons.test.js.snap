// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dragons Dragon component is rendered correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="dragon"
      >
        <img
          alt="Dragon"
          class="dragon-image"
        />
        <div
          class="dragon-details"
        >
          <h3
            class="dragon-name"
          />
          <p
            class="dragon-description"
          />
          <button
            class="dragon-button"
            type="button"
          >
            Reserve Dragon
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="dragon"
    >
      <img
        alt="Dragon"
        class="dragon-image"
      />
      <div
        class="dragon-details"
      >
        <h3
          class="dragon-name"
        />
        <p
          class="dragon-description"
        />
        <button
          class="dragon-button"
          type="button"
        >
          Reserve Dragon
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Dragons Dragons component is rendered correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <ul
          class="dragons-container"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <ul
        class="dragons-container"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
